serve/mlc_serve/engine/metrics.py

```diff
@@ -0,0 +1,43 @@
+from .metrics_labels import *
+from prometheus_client import Counter, Histogram, Gauge
+
+
+class PrometheusMetrics:
+    def __init__(self):
+        self.counters = {}
+        self.histograms = {}
+        self.gauges = {}
+
+        for label in [NUM_CACHE_EVICTONS]:
+            self.counters[label] = Counter(label, label)
+
+        buckets_e2e_lat = (0.5, 2.5, 4.5, 6.5, 8.5, 10.5, 12.5, 14.5, 16.5, 18.5)
+        buckets_ttft = (0.1, 0.3, 0.6, 0.9, 1.2, 1.5, 1.8, 2.1, 2.4, 2.7, 3.0)
+        buckets_batched_prefill_tokens = (500, 1000, 1500, 2000, 2500, 3000, 3500, 4000, 4500, 5000, 6000, 7000, 8000)
+        buckets_batched_decode_tokens = (1, 10, 30, 50, 75, 100, 125, 150, 175, 200, 250, 300)
+
+        for label, buckets in [
+            (E2E_LATENCY, buckets_e2e_lat),
```

Comment:
How do we handle exception in these metrics? Is it included in e2e latency buckets for example?

Reply:
We record a metric only when a request successfully reaches the point in the code where the metric is instrumented. For e2e latency, that's L165 in `stage_engne_worker.py`. So if an exception is raised, the request doesn't reach that point and hence it won't be recoreded.

Reply:
Should we count the number of each type of exceptions raised?

Reply:
That's possible if we have a particular exception in mind. We need to be aware of where that exception could be raised, and add a counter in `except:` block. This is one of follow-up work we can do.